### PR TITLE
[export][unflatten] More strictly respect scope when removing inputs

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127607

Code snippet from TorchTitan (LLaMa):
```
for layer in self.layers.values():
    h = layer(h, self.freqs_cis)
```
`self.freqs_cis` is a buffer of root module (`self`).
It is also an explicit arg in the call signature of original `layer` modules.
If not respecting scope -- `freqs_cis`'s scope only corresponds to root -- `_sink_param` can remove `freqs_cis` from `layer`'s call signature, resulting in runtime error.

There are two fixes in this PR:
1. We filter out the `inputs_to_state` corresponding to the current scope, using existing code that does prefix matching.
2. We delay the removal of param inputs from `call_module` nodes' `args`, till `_sink_param` call on that submodule returns. The return now returns information on which input is actually removed by the submodule, thus more accurate than just doing:
```
    for node in call_module_nodes:
        node.args = tuple(filter(lambda n: n.name not in inputs_to_state, node.args))
```

Before the PR:
Screenshot 2024-05-31 at 1 40 24 AM

After the PR:
Screenshot 2024-05-31 at 1 43 41 AM

### CODE REVIEW

File: torch/export/unflatten.py [hunk 1]
```diff
@@ -337,16 +337,29 @@ def add_to_consts_map(obj_id, node_name, target_name):
...
+        def check_module_inputs(module, scope):
+            if hasattr(module, "graph"):
+                for node in module.graph.nodes:
+                    if (
+                        node.op == "placeholder"
+                        and node.name in inputs_to_state
+                        and any(
+                            fqn.split(".")[: len(scope)] == scope
+                            for fqn in inputs_to_state[node.name]
+                        )  # matching scope to avoid wrong assert
```
Comment (pianpwk) [inline torch/export/unflatten.py:351]: nit: could we add some descriptive comment about the invariant here?

Maybe something like "sink_params() should turn placeholders into get_attr nodes for attributes that are within scope of the current module. We allow attributes to remain as placeholders if they are inputs in the original module signature, meaning they are a parent module's attribute, and therefore out of scope of the current module."

File: torch/export/unflatten.py [hunk 2]
```diff
@@ -1026,32 +1048,47 @@ def _sink_params(
     # Also remove from call_module nodes
     call_module_nodes = filter(lambda n: n.op == "call_module", graph.nodes)
     for node in call_module_nodes:
-        node.args = tuple(filter(lambda n: n.name not in inputs_to_state, node.args))
+        submodule = _recursive_getattr(module, node.target.split("."))
+        if submodule is not None and id(submodule) in module_id_to_inputs_removed:
+            node.args = tuple(
```
Comment (pianpwk) [inline torch/export/unflatten.py:1053]: this is probably the most important line in this PR, so we should add a comment here. Maybe just something simple like "remove placeholder from call_module node arguments, only if we've erased the placeholder node in the corresponding _sink_params() call"

File: torch/export/unflatten.py [hunk 3]
```diff
@@ -1026,32 +1048,47 @@ def _sink_params(
...
+                filter(
+                    lambda n: n.name not in module_id_to_inputs_removed[id(submodule)],
+                    node.args,
+                )
+            )
 
+    # Filter out inputs_to_state corresponding to current scope.
+    inputs_to_state_of_scope: Dict[torch.fx.Node, list[str]] = {}
```
Comment (pianpwk) [inline torch/export/unflatten.py:1061]: Hmm could you actually remind me why we need to split this up into 2 loops now? I don't actually remember...

**Review by pianpwk — APPROVED** [review]
In general looks good to me! Just a couple of nits regarding comments, and one question regarding the 2 for loops. Thanks!

File: torch/export/unflatten.py (hunk 1 above)
Comment (kwen2501) [reply torch/export/unflatten.py:357]: Done, thanks!

File: torch/export/unflatten.py (hunk 2 above)
Comment (kwen2501) [reply torch/export/unflatten.py:1061]: Done, thanks!

File: torch/export/unflatten.py (hunk 3 above)
Comment (kwen2501) [reply torch/export/unflatten.py:1069]: The first loop: `for node in inputs` filters the `inputs_to_state` such that the inputs node **of current scope** is put in `inputs_to_state_of_scope`.
The second loop: `for node, state_name in inputs_to_state_of_scope.items()` actually operate on those nodes.
And yes, I agree that the two loops can be merged too.

**Pull request closed without merging**